cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go

```diff
@@ -458,3 +475,158 @@ func TestValidateFeatureGates(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateArgSelection(t *testing.T) {
+	var tests = []struct {
+		name     string
+		c        *kubeadm.NodeConfiguration
+		expected bool
+	}{
+		{
+			"invalid: DiscoveryToken and DiscoveryFile cannot both be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "https://url/file.conf",
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryToken or DiscoveryFile must be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "",
+				DiscoveryToken: "",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenAPIServers not set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "https://url/file.conf",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"WARNING: using token-based discovery without DiscoveryTokenCACertHashes can be unsafe",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:                         "abcdef.1234567890123456@foobar",
+				DiscoveryTokenUnsafeSkipCAVerification: false,
+				DiscoveryTokenAPIServers:               []string{"192.168.122.100:9898"},
+			},
+			true,
+		},
+		{
+			"WARNING: kubeadm doesn't fully support multiple API Servers yet",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:           "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898", "192.168.122.88:9898"},
+			},
+			true,
+		},
+		{
+			"valid: DiscoveryFile with DiscoveryTokenAPIServers",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:            "https://url/file.conf",
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898"},
+			},
+			true,
+		},
+		{
+			"valid: DiscoveryFile without DiscoveryTokenAPIServers",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile: "https://url/file.conf",
+			},
+			true,
+		},
+		{
+			"valid: using token-based discovery with DiscoveryTokenCACertHashes",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:                         "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers:               []string{"192.168.122.100:9898"},
+				DiscoveryTokenCACertHashes:             []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+				DiscoveryTokenUnsafeSkipCAVerification: false,
+			},
+			true,
+		},
+		{
+			"valid: using token-based discovery with DiscoveryTokenCACertHashe but skip ca verification",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:                         "abcdef.1234567890123456@foobar",
```

Comment:
how can this token be valid with the @? I don't think that's expected?

Reply:
That's an invalid one. Edited.